### ISSUE TITLE
Operator crashing after upgrade

### ISSUE THREAD

**Issue by shanemcd** [ansible/tower-operator#163]
Not exactly sure what's causing this yet, but creating this issue for us to track it:

```
 TASK [Create Database if no database is specified] ********************************
fatal: [localhost]: FAILED! => {"changed": false, "error": 422, "msg": "Failed to apply object: b'{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"StatefulSet.apps \\\\\"awx-postgres\\\\\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than \\'replicas\\', \\'template\\', and \\'updateStrategy\\' are forbidden\",\"reason\":\"Invalid\",\"details\":{\"name\":\"awx-postgres\",\"group\":\"apps\",\"kind\":\"StatefulSet\",\"causes\":[{\"reason\":\"FieldValueForbidden\",\"message\":\"Forbidden: updates to statefulset spec for fields other than \\'replicas\\', \\'template\\', and \\'updateStrategy\\' are forbidden\",\"field\":\"spec\"}]},\"code\":422}\\n'", "reason": "Unprocessable Entity", "status": 422}
```

To fix it I needed to delete and recreate my `awx` CR.

**Comment by tchellomello** [comment]
@shanemcd I´ve noticed this as well while working on the labels issue. It looks is related to changes on the `statefulset`.

**Comment by tchellomello** [comment]
@shanemcd I hit something similar during the upgrade, however, that was during the rollout process
```sh
hase\": \"Running\", \"podIP\": \"10.233.64.159\", \"podIPs\": [{\"ip\": \"10.233.64.159\"}], \"qosClass\": \"Burstable\", \"startTime\": \"2021-04-01T16:16:29Z\"}}]}\n\r\n
TASK [installer : Set the resource pod name as a variable.] ********************\r\n
task path: /opt/ansible/roles/installer/tasks/main.yml:57\nok: [localhost] => {\"ansible_facts\": {\"tower_pod_name\": \"awx-localdb2-postgres-0\"}, \"changed\": false}\n\r\n
TASK [installer : Verify the resource pod name is populated.] ******************\r\n
task path: /opt/ansible/roles/installer/tasks/main.yml:61\nok: [localhost] => {\r\n    \"changed\": false,\r\n    \"msg\": \"All assertions passed\"\r\n}\n\r\n
TASK [installer : Check for pending migrations] ********************************\r\n
task path: /opt/ansible/roles/installer/tasks/main.yml:66\n
fatal: [localhost]: FAILED! => {\"changed\": false, \"msg\": \"Failed to execute on pod awx-localdb2-postgres-0 due to : (0)\\nReason: Handshake status 400 Bad Request\\n\"}\n\r\n
PLAY RECAP *********************************************************************\r\nlocalhost                  : ok=30   changed=1    unreachable=0    failed=1    skipped=24   rescued=0    ignored=0   \r\n\n",
"job":"3175745506366470758","name":"awx-localdb2","namespace":"default","error":"exit status 2","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\n
github.com/operator-framework/operator-sdk/pkg/ansible/runner.(*runner).Run.func1\n\tsrc/github.com/operator-framework/operator-sdk/pkg/ansible/runner/runner.go:239"}
```

On the next cycle, the operator catches up and finished as expected

```yaml
 awx-operator-77957659f4-kwvgc] TASK [installer : Update towerMigratedFromSecret status] ***********************
[awx-operator-77957659f4-kwvgc] task path: /opt/ansible/roles/installer/tasks/update_status.yml:72 
[awx-operator-77957659f4-kwvgc]  
[awx-operator-77957659f4-kwvgc] ------------------------------------------------------------------------------- 
[awx-operator-77957659f4-kwvgc] {"level":"info","ts":1617295426.6505132,"logger":"runner","msg":"Ansible-runner exited successfully","job":"6459015586921615080","name":"awx-localdb2","namespace":"default"} 
[awx-operator-77957659f4-kwvgc]  
[awx-operator-77957659f4-kwvgc] --------------------------- Ansible Task Status Event StdOut  ----------------- 
 awx-operator-77957659f4-kwvgc] 
 awx-operator-77957659f4-kwvgc] PLAY RECAP *********************************************************************
 awx-operator-77957659f4-kwvgc] localhost                  : ok=41   changed=0    unreachable=0    failed=0    skipped=29   rescued=0    ignored=0   
[awx-operator-77957659f4-kwvgc]  
[awx-operator-77957659f4-kwvgc]  
[awx-operator-77957659f4-kwvgc] ------------------------------------------------------------------------------- 
```
So it was not exactly the same issue

**Comment by shanemcd** [comment]
I _think_ we need to use `k8s_info` to check for the presence of the Postgres statefulset before trying to `apply` it?

**Comment by tchellomello** [comment]
> I _think_ we need to use `k8s_info` to check for the presence of the Postgres statefulset before trying to `apply` it?

Yes, we will need to enhance the `statefulset`  apply as the task `Apply Resources`  sometimes fail for that statefulset. I'll work on that.

**Comment by OneHappyForever** [comment]
> To fix it I needed to delete and recreate my awx CR.

Um, what does this mean? You mean the my-awx.yml file or? 

Edit: if you could type a sample command so I can understand what you mean, it would be much appreciated :)

EDIT 2: 
Ok, so CR =  custom resource. But won't deleting it cause the database to lose all the data? Won't it delete the files as well? In this case, how to perform a seemless upgrade without loosing data?

**Comment by shanemcd** [comment]
No, deleting the CR will not delete the PV.

**Comment by OneHappyForever** [comment]
For anyone who need it, this is what I did:

```
minikube kubectl -- delete -f my-awx.yml

minikube kubectl -- apply -f my-awx.yml
```

NOTE: the local port will change. The data is still there.

I still made a snapshot of the VPS just in case.

**Comment by tchellomello** [comment]
OK, I think I made some interesting findings. The problem hit with the `statefulset`  above was related to the labels we added as for `statefulsets`  only a few fields can be modified:  `replicas`, `template` , and `updateStrategy` can be updated. Deleting the `awx`  and having it created again will do the job. 

Maybe we should add notes as `Breaking Changes`  or something similar in the documentation.  I'm still digging more into the `statefulset`  scenario, but I found another issue that is crashing the operator which is also related to #215 (posting there so we can keep things trackable). 

But I think we will need to similar process as verifying if the `statefulset`  changed and if so, then we will need to stop the `awx` deployment, delete the `statefulset`, create again and rollout the new `awx` deployment to be transparent for the user. I will add something along these lines to the #217
